x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/FieldPermissionsDefinition.java

```diff
@@ -92,5 +107,16 @@ public String toString() {
                 + "; exclude=" + Strings.arrayToCommaDelimitedString(excludedFields)
                 + "]";
         }
+
+        @Override
+        public void buildCacheKey(StreamOutput out) throws IOException {
+            out.writeOptionalStringArray(grantedFields);
+            out.writeOptionalStringArray(excludedFields);
+        }
+
+        @Override
+        public int compareTo(FieldGrantExcludeGroup o) {
+            return Integer.compare(hashCode(), o.hashCode());
```

Comment:
What if the hashcodes are equal, but the objects aren't?
Should we be comparing the arrays instead?

Reply:
It could cause less efficient caching because the same group of `FieldGrantExcludeGroup` can be write in different orders. But it will not be a security issue because order along does not change the equality. After all, the exact values (grant and except arrays) from each group are written into the cacheKey. So I chose hashCode for potential faster comparison.
With that being said, I changed it to compare the arrays. It is easier to understand and may not be too slow or slower at all since I never really test.

Reply:
`TreeSet` depends exclusively on `compareTo` it never calls `equals`. From the javadoc:

> a TreeSet instance performs all element comparisons using its compareTo (or compare) method, so two elements that are deemed equal by this method are, from the standpoint of the set, equal.

So returning `0` from `compareTo` for two objects that are not truly equal is an error.